### ISSUE TITLE
PSPNet Cityscapes

### ISSUE THREAD

**Issue by AmeetR** [divamgupta/image-segmentation-keras#78]
Hey,

I've been trying to replicate the published results on cityscapes for pspnet (101). I trained for 15 epochs and after evaluation (I had to change your code a little in order for it to work, I can submit a pull request if you'd like this), the pretrained model scores abysmally (~.02% total IOU) and my model that I trained for 15 epochs got about 27% accuracy... so I'm kind of confused as to why this is. I'm trying to train it for longer now, to see what's going on but yeah I'm pretty confused.

**Comment by AmeetR** [comment]
Additionally, what is the accuracy reported during training, I tried looking through the code for this and couldn't find it. Is it mIOU? If it is, why does that number differ drastically from the one that is generated via eval.py? @divamgupta

**Comment by divamgupta** [comment]
Hi,

The pre-trained model is converted from the caffe model provided by the
authors. Are you sure you are evaluating it correctly?

The accuracy is percentage of pixels correctly classified.

Selecting better hyperparameters / optimiser might improve training results.

**Comment by AmeetR** [comment]
That’s what I’m worried about, I used the logic that you had written as he evaluation function but as I understand it this is not fully tested. The hyperparameter stuff I’m not too worried about rn, at least I want to make sure my eval code is working fine. Thanks, @divamgupta

**Comment by AmeetR** [comment]
def evaluate( model=None , inp_images=None , annotations=None , checkpoints_path=None ):

    #pretrained_model = ks.pretrained.pspnet_101_cityscapes()
    #model = ks.models.pspnet.pspnet_101(20)
    #transfer_weights(pretrained_model, model)
    model = ks.predict.model_from_checkpoint_path(checkpoints_path)
    ious = []
    in_dir = inp_images
    inp_images = os.listdir(inp_images)
    inp_images.sort()
    ann_dir = annotations
    annotations = os.listdir(annotations)
    annotations.sort()
    
    for inp , ann   in zip(inp_images, annotations):
        print(in_dir + inp)
        print(ann)
        pr = ks.predict.predict(model , in_dir + inp )
        gt = ks.predict.get_segmentation_arr( ann_dir + ann , model.n_classes ,  model.output_width , model.output_height  , no_reshape=True)
        plt.imshow(pr)
        print(gt)
        print(gt.shape)
        break
        
        #print(gt.shape)
        gt = gt.argmax(-1)
        #print(gt.shape)
        #print(pr.shape)
        iou = ks.metrics.get_iou( gt , pr , model.n_classes )
        print(iou)
        ious.append( iou )
    ious = np.array( ious )
    print("Class wise IoU "  , np.mean(ious , axis=0 ))
    print("Total  IoU "  ,  np.mean(ious ))



Here's the code that I wrote, most of it is yours I just handled some pre processesing.

**Comment by divamgupta** [comment]
I will look into it. We can take this conservation to email. My id is divam [at] cmu.edu

**Comment by SebastianGer** [comment]
It looks to me like the pspnet._pspnet method is missing the final concatenation with the input features f5, as is indicated in the PSPNet paper in figure 3c. That would explain a bad performance, since the classification is done with only the pooled features. @AmeetR 

Another problem I'm having with the architecture is that it returns a segmentation map that is 1/8 of width and height of the input. The architecture itself already contains a resize by factor 8 at the end (which then outputs a segmentation of size 1/8 * (width,height)), then the predict.predict method applies another resize to achieve the original size. That also means that we only have one prediction for each 64x64 pixel area.

Judging from [this thread](https://github.com/hszhao/PSPNet/issues/101) on the original implementation, the resizing in the predict.predict method shouldn't be necessary. However, I'm not sure, where the problem occurs here, or if I'm just misunderstandings something. @divamgupta
